src/nbody/barnes_hut.py

```diff
@@ -0,0 +1,373 @@
+
+        Returns:
+            np.ndarray: The acceleration vector.
+        """
+        accel = np.zeros(3)
+
+        stack = [self.root]
+
+        while stack:
+            node = stack.pop()
+            dist = np.linalg.norm(node.center_of_mass - other.position)
+            # accept any cube with l/D < theta
+            length = node.length
+            if (
+                dist == 0
+            ):  # if a cube happens to have center of mass at the point, examine children, unless
+                # it is a point cube
+                if node.children:
+                    stack.extend(node.children)
+                else:
+                    continue
+            if length / dist < theta:
+                dir = (node.center_of_mass - other.position) / dist
+                accel += dir * G * node.mass / (dist**2)
+            else:
+                if node.children:
+                    stack.extend(node.children)
+                else:  # point cubes should always be accepted
+                    dir = (node.center_of_mass - other.position) / dist
+                    accel += dir * G * node.mass / (dist**2)
+        return accel
+
+    def __str__(self):
+        return f"Tree\n{self.root!s}"
+
+
+class BarnesHutPropagator:
+    """
+    Simulate an orbital system using the Barnes-Hut N logN algorithm for force computation.
+    """
+
+    def __init__(
+        self,
+        bodies: list[Body],
+        params: dict | None,
+        output_file: str = "OUTPUT_FILES/traces.csv",
+    ):
+        """
+        Arguments:
+            bodies (list[Body]): The bodies in the system to be simulated.
+            params (dict): Simulation parameters. Supported keys are "dt", "t_max", "theta", and "G"
+                (defaults to 1, 100, 1, and 6.6743e-11, resp.)
+            output_file (str): The filename that traces are written to. Defaults to `OUTPUT_FILES/traces.csv`.
+        """
+        if params is None:
+            params = {}
+        dt = params.get("dt", 1.0)
+        t_max = params.get("t_max", 100)
+        theta = params.get("theta", 1)
+        G = params.get("G", 6.6743e-11)
+
+        path = Path(output_file)
+        path.parent.mkdir(exist_ok=True, parents=True)
+        with Path.open(path, "w") as f:
+            writer = csv.writer(f)
+            # The header line
+            header = ["Iteration", "t"]
+            for i in range(1, len(bodies) + 1):
+                header.extend(
+                    [
+                        f"Body {i} position (x)",
+                        f"Body {i} position (y)",
+                        f"Body {i} position (z)",
+                        f"Body {i} velocity (x)",
+                        f"Body {i} velocity (y)",
+                        f"Body {i} velocity (z)",
+                        f"Body {i} acceleration (x)",
+                        f"Body {i} acceleration (y)",
+                        f"Body {i} acceleration (z)",
+                    ]
```

Comment:
```suggestion
            header = ["Iteration", "t"] + [
            f"Body {i} {prop} ({axis})"
            for i in range(1, len(bodies) + 1)
            for prop in ("position", "velocity", "acceleration")
            for axis in ("x", "y", "z")
            ]
```

This or a python-ey for-element like it should work as a more compact expression for the header.